New resource to work with Windows security identifiers (SIDs)

james-stocks opened the pull request:
We're currently making use of this resource as a custom resource in our profiles; I would like to propose adding it into Inspec.

Example control that this new resource helps with when testing `security_policy`:

```
permitted_users = ['S-1-5-9', security_identifier(group: 'Administrators').sid, security_identifier(user: 'Administrator').sid]
  filtered_principals = security_policy.SeNetworkLogonRight - permitted_users
  filtered_principals.each do |suspect_principal|
    describe(suspect_principal) do
      it { should_not match /.*/ }
    end
  end
```

_Why bother with this resource instead of stating the SIDs as literal strings?_
For one, some SIDs are dynamic (include the domain ID). Secondly, some automated compliance content will use user and group names when, in practice, the SID is needed. To automatically consume and execute this content, we need an automatic way to convert an entity name to a SID.

_Why bother needing the user to specify user, group or unspecified_
It is possible to create users and groups and other entity types on Windows with the same name. It may cause inaccurate results and leave a slight gap in security if the name of the entity being tested is ambiguous.
The `unspecified` type is needed because some automated compliance content will name entities that should be tested without indicating what type of entity they are. To keep this automated, we need a way to try to determine a SID when we're given an entity name but not the entity type. Ideally, `unspecified` should not be used.

Commit:
Resource for a Windows Security Identifier (SID)

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Integration tests for security_identifier resource

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Address rubocop violations

Signed-off-by: James Stocks <jstocks@chef.io>

clintoncwolfe commented on test/integration/default/controls/security_identifier_spec.rb:
```diff
@@ -0,0 +1,32 @@
+  its('sid') { should match %r{S-1-5-21.+-500} }
+end
+
+# If not appropriate, sid just returns the name back
```
Hrm, this would be atypical behavior compared to other InSpec core resources. When a resource performs a search, the usual way of describing a "miss" is:

* set `.exists?` to false
* set all properties to `nil`

clintoncwolfe commented on test/integration/default/controls/security_identifier_spec.rb:
```diff
@@ -0,0 +1,32 @@
+  STDERR.puts "\033[1;33mTODO: Not running #{__FILE__} because we are not on Windows.\033[0m"
+  return
+end
+
```
I think you should have an `.exist?` property.  That is more typical for resources.

james-stocks replied:
When you suggest an `.exist?` property I'm assuming this is tested with `it { should exist }` , right?

clintoncwolfe replied:
Yes, that's right.

clintoncwolfe commented on test/integration/default/controls/security_identifier_spec.rb:
```diff
@@ -0,0 +1,32 @@
+  its('sid') { should match %r{S-1-5-21.+-500} }
+end
+
+describe security_identifier({ unspecified: 'Administrator' }) do
```
Great start on the tests!

clintoncwolfe commented on test/unit/resources/security_identifier_test.rb:
```diff
@@ -0,0 +1,11 @@
+  it 'instantiates' do
+    resource = MockLoader.new(:windows).load_resource('security_identifier', { user => 'Administrator' })
+    _(resource).wont_be_nil
+  end
```
Since you accept a variety of search keys (user, group, unspecified) you should probably test that all 3 of those are accepted.  Then you should try passing an unrecognized search key, and verify that an ArgumentError or similar is thrown.

james-stocks replied:
Have added tests for these

clintoncwolfe commented on test/unit/resources/security_identifier_test.rb:
```diff
@@ -0,0 +1,11 @@
+require 'inspec/resource'
+
+describe 'Inspec::Resources::SecurityIdentifier' do
+  it 'instantiates' do
```
I would prefer that this test be expanded to also test actual searches, and their results.  For example, you could have a mock that knows about a user named Alice, and verify that you can find Alice.  We're not talking to an actual Windows machine here, but the goal would be to show that you can interpret the output of the machine properly; and you can add data (like Alice) that would not be on an integration test fixture.

clintoncwolfe commented on lib/resources/security_identifier.rb:
```diff
@@ -0,0 +1,66 @@
+  class SecurityIdentifier < Inspec.resource(1)
+    name 'security_identifier'
+    supports platform: 'windows'
+    desc 'Resource that returns a Security Identifier for a given entity name in Windows. Because different entities can have the same name (e.g. a user and group can both be called \'devops\') the resource requires the type of the entity (:user, :group) to be stated to avoid an ambiguous test. If the entity type is unknown (e.g. when working with SCAP content that names an entity but does not de ...
```
This should be moved to docs/resources/security_identifier.md.erb , and expanded to have the section headers typical of a resource doc.  Thanks!

You can leave a one-line description here.

clintoncwolfe commented on lib/resources/security_identifier.rb:
```diff
@@ -0,0 +1,66 @@
+    name 'security_identifier'
+    supports platform: 'windows'
+    desc 'Resource that returns a Security Identifier for a given entity name in Windows. Because different entities can have the same name (e.g. a user and group can both be called \'devops\') the resource requires the type of the entity (:user, :group) to be stated to avoid an ambiguous test. If the entity type is unknown (e.g. when working with SCAP content that names an entity but does not de ...
+    example "
```
Consider using heredoc markers, which will automatically escape quotes.  That preserves the "copy-paste" ability that a lot of users want.

```
example <<~EOD
  describe security_policy do
    its("SeRemoteInteractiveLogonRight") { should_not include security_identifier(group: 'Guests') }
   end
EOD
```

clintoncwolfe replied:
While I understand this is primarily a data lookup resource, you should still show it in use alone.  Having two examples - one in which it is used alone, and one 'advanced example' in which it is used in conjunction with a `security_policy` would be clearer to show what this resource does.

clintoncwolfe commented on lib/resources/security_identifier.rb:
```diff
@@ -0,0 +1,66 @@
+
+    def initialize(opts = {})
+      supported_opt_keys = [:user, :group, :unspecified]
+      raise "Invalid security_identifier param '#{opts}'. Please pass a hash with these supported keys: #{supported_opt_keys}" unless opts.respond_to?(:keys)
```
Please raise ArgumentError exceptions when an arg is invalid.

clintoncwolfe commented on lib/resources/security_identifier.rb:
```diff
@@ -0,0 +1,66 @@
+
+    def sid
+      fetch_sids unless @sids
+      @sids[@name] || @name
```
If you can't find something, return nil, and have an `exist?` property.  That's the usual interface.

clintoncwolfe commented on lib/resources/security_identifier.rb:
```diff
@@ -0,0 +1,66 @@
+    end
+
+    def wmi_results_array(query)
+      inspec.command(query).stdout.strip.split("\r\n\r\n")[1..-1].map { |entry| entry.split(',') }
```
Do you need the `\r`?  `\n` is typically the OS-Specific newline, so it should handle that on its own.

clintoncwolfe replied:
It would be kinder to your fellow developers to leave a comment with an example of the CLI data you are parsing, and the data structure you intend to return.

clintoncwolfe commented on lib/resources/security_identifier.rb:
```diff
@@ -0,0 +1,66 @@
+      sid_data.each { |sid| @sids[sid[1]] = sid[2] }
+    end
+
+    def wmi_results_array(query)
```
You can probably drop the suffix `array`.

clintoncwolfe commented on lib/resources/security_identifier.rb:
```diff
@@ -0,0 +1,66 @@
+      @sids = {}
+      case @type
+      when :group
+        sid_data = wmi_results_array("wmic group where 'Name=\"#{@name}\"' get Name\",\"SID /format:csv")
```
It looks like only the `group`/`useraccount` part actually varies in each of these invocations.  Consider moving the construction of the command line into `wmi_results_array`, so you only are obliged to build the complex command line in one place, not 4.

clintoncwolfe reviewed (changes requested):
Thanks for the contribution!  I agree, this would be great data lookup resource to have, and I think you have made a great start.  

I'd ask that it handle search misses differently, move the docs to a markdown file, and improve test coverage.  It's a great start, and can easily be matured into a strong resource.  Thanks again!

PS - Looks like you have a failing unit test - see https://travis-ci.org/inspec/inspec/jobs/430099348#L1237

Commit:
Improve security_identifier from PR feedback

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Update security_identifier tests

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Improve security_identifier unit tests

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Fix unit tests fpr security_identifier resource

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
More security_identifier unit tests

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Add docs page for security_identifier resource

Signed-off-by: James Stocks <jstocks@chef.io>

james-stocks commented on docs/resources/security_identifier.md.erb:
```diff
@@ -0,0 +1,63 @@
+
+### Version
+
+This resource first became available in vX.Y.Z of InSpec.
```
Can I just put the next version number for Inspec into this line?

clintoncwolfe replied:
Simply omit the Version section for now.

clintoncwolfe commented on docs/resources/security_identifier.md.erb:
```diff
@@ -0,0 +1,63 @@
+where
+
+* `group:` specifies that `'Everyone'` should be a group. `user:` can be used to specify a user account.
+** It is necessary to declare the type of the entity because Windows allows users, groups and other entities to share names. If you really need to not specify the type, `unsupported:` can be used. This will attempt to match the name to a group and then a useraccount. This may take longer to execute and comes with the risk of Inspec matching the name to an unintended type of entity. 
```
I think you mean 'unspecified', not 'unsupported'

clintoncwolfe replied:
When referred to as a product, we try to refer to it as InSpec (note caps).  As a CLI or gem, use 'inspec'.

clintoncwolfe commented on docs/resources/security_identifier.md.erb:
```diff
@@ -0,0 +1,63 @@
+
+This resource first became available in vX.Y.Z of InSpec.
+
+## Syntax
```
Prefer 'Resource Parameters' as a header instead of 'Syntax'.  We're inconsistent about this :/

clintoncwolfe commented on docs/resources/security_identifier.md.erb:
```diff
@@ -0,0 +1,63 @@
+### Use in conjunction with the security_policy resource to specify the entity to test for in the audit policy
+
+    describe security_policy do
+      its(\"SeRemoteInteractiveLogonRight\") { should_not include security_identifier(group: 'Guests') }
```
I don't think the escapes on the quotes are needed here; and will be syntax errors if someone copy-pastes this.

clintoncwolfe commented on docs/resources/security_identifier.md.erb:
```diff
@@ -0,0 +1,63 @@
+
+<br>
+
+## Matchers
```
You should list `exist` here.

james-stocks replied:
Added

clintoncwolfe commented on docs/resources/security_identifier.md.erb:
```diff
@@ -0,0 +1,63 @@
+
+<br>
+
+## Matchers
```
You are missing a section, 'Properties', in which you explain anything you can pass to `its`.  At least `sid`.

james-stocks replied:
Thanks, added

clintoncwolfe commented on lib/resources/security_identifier.rb:
```diff
@@ -0,0 +1,84 @@
+      #     inspec> command("wmic useraccount where 'Name=\"Administrator\"' get Name\",\"SID /format:csv").stdout
+      #     => "\r\n\r\nNode,Name,SID\r\n\r\nComputer1,Administrator,S-1-5-21-650485088-1194226989-968533923-500\r\n\r\n"
+      # Remove the \r characters, split on \n\n, ignore the CSV header row
+      inspec.command(query).stdout.strip.tr("\r", '').split("\n\n")[1..-1].map { |entry| entry.split(',') }
```
Very good comments!

Commit:
Fix issues with documentation

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Improve docs

Link to Microsoft reference page, and use their term 'trustee' instead of 'entity' where applicable.

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Change exists to exist

Signed-off-by: James Stocks <jstocks@chef.io>

jquick reviewed (approved):
Looks good. Thanks @james-stocks

jquick committed:
Test appveyor file changes.

Signed-off-by: Jared Quick <jquick@chef.io>
